ecosystem/cache/handlers.go

```diff
@@ -120,6 +147,37 @@ func (s *RelayerCacheServer) getRelayInner(ctx context.Context, relayCacheGet *p
 	return nil, HashMismatchError
 }
 
+// this method tries to set the seen block a few times until it succeeds. it will try to make sure its value is set
+// to prevent race conditions when we have a few writes in the same time with older values we will try to set our value eventually
+// if its the newest seen block
+func (s *RelayerCacheServer) setSeenBlockOnSharedStateMode(chainId, sharedStateId string, seenBlock int64) {
+	if sharedStateId == "" {
+		return
+	}
+	existingSeenBlock := s.getSeenBlockForSharedStateMode(chainId, sharedStateId)
+	// validate we have a newer block than the existing stored in the db.
+	if existingSeenBlock <= seenBlock { // refreshes state even if its equal
+		key := latestBlockKey(chainId, sharedStateId)
+		// for seen block we expire the entry after one hour otherwise this user will stay in the db for ever
+		s.CacheServer.finalizedCache.SetWithTTL(key, seenBlock, 0, s.CacheServer.ExpirationFinalized)
+		// a validation routine to make sure we don't have a race for seen block rewrites as there are concurrent writes.
+		// this will be solved once we implement a db with a queue but for now its a good enough solution.
+		go func() {
```

Comment:
can we do both retry functions here and on set latest in the same function (get the write as a callback), later we will replace it and we don;t want to do it in many places